src/librustc_mir/build/expr/as_rvalue.rs

```diff
@@ -517,21 +514,19 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
             }
             Place {
                 ref base,
-                projection: Some(box Projection {
-                    base: ref base_proj,
-                    elem: ProjectionElem::Field(upvar_index, _),
-                }),
+                projection: box [.., ProjectionElem::Field(upvar_index, _)],
             }
             | Place {
                 ref base,
-                projection: Some(box Projection {
-                    base: Some(box Projection {
-                        base: ref base_proj,
-                        elem: ProjectionElem::Field(upvar_index, _),
-                    }),
-                    elem: ProjectionElem::Deref,
-                }),
+                projection: box [.., ProjectionElem::Field(upvar_index, _), ProjectionElem::Deref],
             } => {
+                let base_proj =
+                    if let ProjectionElem::Deref = arg_place.projection[arg_place.projection.len() - 1] {
```

Comment:
You have already proven that the tail element is `::Deref` here; the `else` branch should be dead code.

Reply:
We are matching with a `|` above. So that if is the way I have to know what was the pattern that actually matched. This is another case where slice_patterns would help a lot. Waiting for the beta release :).

Reply:
ah; didn't notice that, my bad :)